### PR TITLE
Separate activation to layer @open sesame 06/18 15:46

### LINKED ISSUE #153: Decouple activation to `layer`
Might be related to #152.
### AS-IS
Current implementation of `fc_layer` has activation function and loss function coupled.

### TO-BE
IMO, decoupling the activation layer as well as loss function(as @kparichay noted) is also needed to increase modularity.

for example in Keras, because activation functionality is decoupled, putting `bn_layer` between `fc` and `activation` can be done easily.

```python
model.add(Dense(64))
model.add(BatchNormalization())
model.add(Activation('tanh'))
```

still activation can be used as an argument of `fc_layer` possible because `fc_layer` can have `activation_layer`

```python
model.add(Dense(64, activation="relu" ))
```

### Proposal
- Add `activation_layer`(or equivalent) and extract the functionality.
- Change `fc_layer` to have `activation_layer`

@jijoongmoon Could you confirm if this change is okay?

### PR DESCRIPTION
**Pull request by zhoonit**
**Changes proposed in this PR:**
- add activation_layer.[h|cpp] (~Partly~ implemented)
- add test to activation_layer(#190)
- add activation_layer when ini find activation property(~planned to~ moved to another PR)
- change cost calculation to be independent from actiType(maybe in another PR, TBD)
    - add `COST_ENTROPY_SIGMOID` to `CostType`
    - add `COST_ENTROPY_SOFTMAX`
- move `activation_type` to activation_layer (NYI, maybe in another PR)
- remove `activation`, `activation_prime` from Layer(NYI, maybe in another PR)

see also #153, #152 

**Self evaluation:**
1. Build test: [ ]Passed [ ]Failed [X]Skipped
2. Run test: [ ]Passed [ ]Failed [X]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

### CODE REVIEW

File: nntrainer/src/activation_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+/**
+ * @brief     set Property of layer
+ * @param[in] values values of property
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+ */
+int ActivationLayer::setProperty(std::vector<std::string> values) {
+  int status = ML_ERROR_NONE;
+
+  if (values.size() > 1 || values.empty()) {
```
Comment (dongju-chae) [inline nntrainer/src/activation_layer.cpp:123]: Minor comment: I think `values.size() != 1` is enough

Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp:123]: Right, thanks!

**Comment by zhoonit** [comment]
v2:
** Changes proposed in this PR: **

 - add activation_layer.[h|cpp] (implementation done and ready to be reviewed)

File: nntrainer/include/activation_layer.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0-only
+ *
+ * @file	activation_layer.h
+ * @date	17 June 2020
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Jihoon Lee <jhoon.it.lee@samsung.com>
```
Comment (jijoongmoon) [inline nntrainer/include/activation_layer.h:9]: remove tab here?

File: nntrainer/include/activation_layer.h [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ * @bug		No known bugs except for NYI items
+ * @brief	This is Activation Layer Class for Neural Network
+ *
+ */
+
+#ifndef __ACTIVATION_LAYER_H_
+#define __ACTIVATION_LAYER_H_
```
Comment (jijoongmoon) [inline nntrainer/include/activation_layer.h:16]: better to use ``` __ACTIVATION_LAYER_H__```

File: nntrainer/include/activation_layer.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+   *
+   * @param[in] std::function<Tensor(Tensor const &)> activation_fn activation
+   *            function to be used
+   * @param[in] std::function<Tensor(Tensor const &)> activation_prime_fn
+   *            activation_prime_function to be used
+   */
+  void setActivation(
+    std::function<Tensor(Tensor const &)> const &activation_fn,
+    std::function<Tensor(Tensor const &)> const &activation_prime_fn);
+
```
Comment (jijoongmoon) [inline nntrainer/include/activation_layer.h:100]: It's good idea.

File: nntrainer/src/activation_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0-only
+ *
+ * @file	activation_layer.cpp
+ * @date	17 June 2020
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Jihoon Lee <jhoon.it.lee@samsung.com>
```
Comment (jijoongmoon) [inline nntrainer/src/activation_layer.cpp:9]: remove tab

File: nntrainer/src/activation_layer.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+#include <util_func.h>
+#include <vector>
+
+namespace nntrainer {
+
+/**
+ * @class   Activation Layer
+ * @brief   Activation Layer
+ */
+
```
Comment (jijoongmoon) [inline nntrainer/src/activation_layer.cpp:33]: Remove this comment?

File: nntrainer/src/activation_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+/**
+ * @brief     Initialize the layer
+ *
+ * @param[in] last last layer
+ * @retval #ML_ERROR_NONE Successful.
+ * @retval #ML_ERROR_INVALID_PARAMETER invalid parameter.
+ */
+int ActivationLayer::initialize(bool last) {
+  this->type = LAYER_ACTIVATION;
+  return ML_ERROR_NONE;
```
Comment (jijoongmoon) [inline nntrainer/src/activation_layer.cpp:43]: this can be done at the construct of Activation Layer.

File: nntrainer/src/activation_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+void ActivationLayer::read(std::ifstream &file){
+  /* noop */
```
Comment (jijoongmoon) [inline nntrainer/src/activation_layer.cpp:47]: maybe just using header?

File: nntrainer/src/activation_layer.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+  std::shared_ptr<ActivationLayer> from =
+    std::static_pointer_cast<ActivationLayer>(l);
+  this->input.copy(from->input);
+  this->hidden.copy(from->hidden);
+  this->activation_type = from->activation_type;
+
+};
+
+void ActivationLayer::setActivation(
+  std::function<Tensor(Tensor const &)> const &activation_fn,
```
Comment (jijoongmoon) [inline nntrainer/src/activation_layer.cpp:81]: I think it might be better to have error handling. ``` return ML_ERROR_NONE ```

File: nntrainer/src/activation_layer.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+  _act_prime_fn = [activation_prime_fn](Tensor const &t) { return t.apply(activation_prime_fn); };
+}
+
+/**
+ * @brief setActivation by preset actiType
+ *
+ * @param[in] ActiType actiType actiType to be set
+ */
+void ActivationLayer::setActivation(ActiType acti_type) {
+  this->activation_type = acti_type;
```
Comment (kparichay) [inline nntrainer/src/activation_layer.cpp:100]: How about setting this in the end? its better not to modify this value is we are going to throw an exception.

File: nntrainer/src/activation_layer.cpp (hunk 7 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp:43]: Activation_layer's  `Forward` and `Backward` does not need to have `input_dim` and `output_dim` internally as this is always the same as input. Would this be used outside?

File: nntrainer/include/activation_layer.h (hunk 2 above)
Comment (zhoonit) [reply nntrainer/include/activation_layer.h]: oops.. fixed! Thank you!

File: nntrainer/src/activation_layer.cpp (hunk 5 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp]: Fixed! Thank you :)

File: nntrainer/src/activation_layer.cpp (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/src/activation_layer.cpp]: Yes. I is used by initialize of neural network.

File: nntrainer/include/activation_layer.h (hunk 3 above)
Comment (zhoonit) [reply nntrainer/include/activation_layer.h]: the last underscore ran away, replaced with new one. Thanks

File: nntrainer/src/activation_layer.cpp (hunk 6 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp]: Fixed.. thanks!

File: nntrainer/src/activation_layer.cpp (hunk 8 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp]: Fixed. PTAL. Thank you!

File: nntrainer/src/activation_layer.cpp (hunk 9 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp:77]: I changed to return ML_ERROR_NONE. Thank you!

File: nntrainer/src/activation_layer.cpp (hunk 10 above)
Comment (zhoonit) [reply nntrainer/src/activation_layer.cpp]: Thanks for the suggestion. I changed it!

**Comment by kparichay** [comment]
This might not be the right place to ask but I want to bring this to your notice - 
Some layers like `fc` (and `conv` in future) change their calculation for `backwarding` if they are the last layer. Adding the activation layer after them could change those semantics. Please keep a check on this when you update `neuralnet.cpp` where network is created with the activation layer.

**Comment by zhoonit** [comment]
> This PR looks good to me
> However, I want to bring this to your notice that some layers like `fc` (and `conv` in future) change their calculation for `backwarding` if they are the last layer. Adding the activation layer after them could change those semantics. The concept of the `last layer` will need some updation when you update `neuralnet.cpp` with the activation layer.
> 

Acked. thanks for noticing the issue.
